File: schema/prompt.go
package schema

import (
	"encoding/json"
	"regexp"
	"strconv"
	"strings"
)

// ============================================================
// Prompt Schema Types
// ============================================================

// PromptSchema is a compact, LLM-friendly projection of a
// DatasourceConfigSchema. It strips UI hints, groups, storage
// mappings, and other rendering/internal concerns.
type PromptSchema struct {
	PluginType string        `json:"pluginType"`
	PluginName string        `json:"pluginName"`
	Fields     []PromptField `json:"fields"`
}

// PromptField is a single field projected for LLM consumption.
type PromptField struct {
	ID            string         `json:"id"`
	Path          string         `json:"path"`
	Type          ValueType      `json:"type"`
	Label         string         `json:"label,omitempty"`
	SemanticType  SemanticType   `json:"semanticType,omitempty"`
	Description   string         `json:"description,omitempty"`
	Required      bool           `json:"required,omitempty"`
	RequiredWhen  string         `json:"requiredWhen,omitempty"`
	DependsOn     string         `json:"dependsOn,omitempty"`
	DefaultValue  any            `json:"defaultValue,omitempty"`
	AllowedValues []any          `json:"allowedValues,omitempty"`
	Pattern       string         `json:"pattern,omitempty"`
	Range         *PromptRange   `json:"range,omitempty"`
	Items         []PromptField  `json:"items,omitempty"`
	Options       []PromptOption `json:"options,omitempty"`
}

// PromptRange is a numeric bound constraint.
type PromptRange struct {
	Min *float64 `json:"min,omitempty"`
	Max *float64 `json:"max,omitempty"`
}

// PromptOption is a virtual selector option with side-effects.
type PromptOption struct {
	Value any            `json:"value"`
	Label string         `json:"label"`
	Sets  map[string]any `json:"sets"`
}

// ============================================================
// Projection
// ============================================================

// ToPromptSchema projects a full DatasourceConfigSchema into a
// compact, LLM-friendly PromptSchema.
func ToPromptSchema(s *DatasourceConfigSchema) PromptSchema {
	var fields []PromptField
	for i := range s.Fields {
		if isManagedField(&s.Fields[i]) {
			continue
		}
		fields = append(fields, projectField(&s.Fields[i]))
	}
	return PromptSchema{
		PluginType: s.PluginType,
		PluginName: s.PluginName,
		Fields:     fields,
	}
}

// ToPromptString serializes a DatasourceConfigSchema into a compact
// JSON string suitable for embedding in an LLM prompt (tool/function calling).
func ToPromptString(s *DatasourceConfigSchema) (string, error) {
	ps := ToPromptSchema(s)
	data, err := json.MarshalIndent(ps, "", "  ")
	if err != nil {
		return "", err
	}
	return string(data), nil
}

// ToPromptText renders a DatasourceConfigSchema into human-readable
// text suitable for embedding in an LLM system/user prompt.
// More token-efficient and easier for LLMs to reason about than JSON.
func ToPromptText(s *DatasourceConfigSchema) string {
	ps := ToPromptSchema(s)
	var b strings.Builder

	b.WriteString(ps.PluginName)
	b.WriteString(" (pluginType: ")
	b.WriteString(ps.PluginType)
	b.WriteString(")\n\nFields:\n")

	for i := range ps.Fields {
		renderField(&b, &ps.Fields[i], "")
	}

	return b.String()
}

func renderField(b *strings.Builder, f *PromptField, indent string) {
	// Main line
	name := f.Label
	if name == "" {
		name = f.ID
	}

	b.WriteString(indent)
	b.WriteString("- ")
	b.WriteString(name)

	if strings.Contains(f.Path, ".") {
		b.WriteString(" (")
		b.WriteString(f.Path)
		b.WriteString(")")
	}

	b.WriteString(" [")
	b.WriteString(string(f.Type))
	if f.SemanticType != "" {
		b.WriteString(", ")
		b.WriteString(string(f.SemanticType))
	}
	b.WriteString("]")

	if f.Required {
		b.WriteString(" REQUIRED")
	}
	if f.DefaultValue != nil {
		b.WriteString(" default: ")
		b.WriteString(formatValue(f.DefaultValue))
	}
	if f.Description != "" {
		b.WriteString(" — ")
		b.WriteString(f.Description)
	}
	b.WriteString("\n")

	sub := indent + "  "

	if f.RequiredWhen != "" {
		b.WriteString(sub)
		b.WriteString("Required when: ")
		b.WriteString(f.RequiredWhen)
		b.WriteString("\n")
	}
	if f.DependsOn != "" {
		b.WriteString(sub)
		b.WriteString("Visible when: ")
		b.WriteString(f.DependsOn)
		b.WriteString("\n")
	}
	if len(f.AllowedValues) > 0 && len(f.Options) == 0 {
		b.WriteString(sub)
		b.WriteString("Allowed: ")
		for i, v := range f.AllowedValues {
			if i > 0 {
				b.WriteString(", ")
			}
			b.WriteString(formatValue(v))
		}
		b.WriteString("\n")
	}
	if f.Pattern != "" {
		b.WriteString(sub)
		b.WriteString("Pattern: ")
		b.WriteString(f.Pattern)
		b.WriteString("\n")
	}
	if f.Range != nil {
		b.WriteString(sub)
		b.WriteString("Range: ")
		parts := []string{}
		if f.Range.Min != nil {
			parts = append(parts, "min: "+strconv.FormatFloat(*f.Range.Min, 'f', -1, 64))
		}
		if f.Range.Max != nil {
			parts = append(parts, "max: "+strconv.FormatFloat(*f.Range.Max, 'f', -1, 64))
		}
		b.WriteString(strings.Join(parts, ", "))
		b.WriteString("\n")
	}

	if len(f.Options) > 0 {
		b.WriteString(sub)
		b.WriteString("Options:\n")
		for _, opt := range f.Options {
			b.WriteString(sub)
			b.WriteString("  ")
			b.WriteString(formatValue(opt.Value))
			b.WriteString(" (")
			b.WriteString(opt.Label)
			b.WriteString(") → sets ")
			first := true
			for k, v := range opt.Sets {
				if !first {
					b.WriteString(", ")
				}
				b.WriteString(k)
				b.WriteString("=")
				b.WriteString(formatValue(v))
				first = false
			}
			b.WriteString("\n")
		}
	}

	if len(f.Items) > 0 {
		b.WriteString(sub)
		b.WriteString("Item fields:\n")
		for i := range f.Items {
			renderField(b, &f.Items[i], sub+"  ")
		}
	}
}

func formatValue(v any) string {
	switch val := v.(type) {
	case string:
		return `"` + val + `"`
	case bool:
		if val {
			return "true"
		}
		return "false"
	case float64:
		return strconv.FormatFloat(val, 'f', -1, 64)
	case int:
		return strconv.Itoa(val)
	case nil:
		return "null"
	default:
		data, _ := json.Marshal(val)
		return string(data)
	}
}

// ============================================================
// Internals
// ============================================================

// isManagedField returns true if the field is tagged "managed-by:*",
// meaning it is driven by effects and should be hidden from the LLM.
func isManagedField(f *ConfigField) bool {
	for _, tag := range f.Tags {
		if strings.HasPrefix(tag, "managed-by:") {
			return true
		}
	}
	return false
}

func projectField(f *ConfigField) PromptField {
	pf := PromptField{
		ID:   f.ID,
		Path: f.Path(),
		Type: f.ValueType,
	}

	if f.Label != "" {
		pf.Label = f.Label
	}
	if f.SemanticType != "" {
		pf.SemanticType = f.SemanticType
	}
	if f.Description != "" {
		pf.Description = f.Description
	}
	if f.Required {
		pf.Required = true
	}
	if f.RequiredWhen != "" {
		pf.RequiredWhen = f.RequiredWhen
	}
	if f.DependsOn != "" {
		pf.DependsOn = f.DependsOn
	}
	if f.DefaultValue != nil {
		pf.DefaultValue = f.DefaultValue
	}

	// Flatten validations
	for _, v := range f.Validations {
		switch v.Type {
		case AllowedValuesValidation:
			pf.AllowedValues = v.Values
		case PatternValidation:
			pf.Pattern = v.Pattern
		case RangeValidation:
			pf.Range = &PromptRange{Min: v.Min, Max: v.Max}
		}
	}

	// Flatten effects into options
	if len(f.Effects) > 0 {
		labelMap := buildLabelMap(f)
		for _, eff := range f.Effects {
			val := ExtractLiteralFromWhen(eff.When)
			label := ""
			if val != nil {
				if l, ok := labelMap[toString(val)]; ok {
					label = l
				} else {
					label = toString(val)
				}
			} else {
				label = eff.When
			}
			optVal := val
			if optVal == nil {
				optVal = eff.When
			}
			pf.Options = append(pf.Options, PromptOption{
				Value: optVal,
				Label: label,
				Sets:  eff.Set,
			})
		}
	}

	// Recurse into array/map item fields
	if f.Item != nil && len(f.Item.Fields) > 0 {
		for i := range f.Item.Fields {
			pf.Items = append(pf.Items, projectField(&f.Item.Fields[i]))
		}
	}

	return pf
}

func buildLabelMap(f *ConfigField) map[string]string {
	m := make(map[string]string)
	if f.UI != nil {
		for _, opt := range f.UI.Options {
			m[toString(opt.Value)] = opt.Label
		}
	}
	return m
}

func toString(v any) string {
	switch val := v.(type) {
	case string:
		return val
	case bool:
		if val {
			return "true"
		}
		return "false"
	case float64:
		return strconv.FormatFloat(val, 'f', -1, 64)
	case int:
		return strconv.Itoa(val)
	default:
		return ""
	}
}

// ExtractLiteralFromWhen extracts a literal value from a simple
// "value == '...'" expression. Returns nil for complex expressions.
var (
	reSingleQuote = regexp.MustCompile(`^value\s*==\s*'([^']*)'$`)
	reDoubleQuote = regexp.MustCompile(`^value\s*==\s*"([^"]*)"$`)
	reBool        = regexp.MustCompile(`^value\s*==\s*(true|false)$`)
	reNumber      = regexp.MustCompile(`^value\s*==\s*(-?\d+(?:\.\d+)?)$`)
)

func ExtractLiteralFromWhen(when string) any {
	if m := reSingleQuote.FindStringSubmatch(when); m != nil {
		return m[1]
	}
	if m := reDoubleQuote.FindStringSubmatch(when); m != nil {
		return m[1]
	}
	if m := reBool.FindStringSubmatch(when); m != nil {
		return m[1] == "true"
	}
	if m := reNumber.FindStringSubmatch(when); m != nil {
		n, err := strconv.ParseFloat(m[1], 64)
		if err == nil {
			return n
		}
	}
	return nil
}

File: schema/prompt.ts
import type {
    DatasourceConfigSchema,
    ConfigField,
    ValueType,
    SemanticType,
    FieldEffect,
} from "./schema"

// ============================================================
// Prompt Schema Types
// ============================================================

export interface PromptSchema {
    pluginType: string
    pluginName: string
    fields: PromptField[]
}

export interface PromptField {
    id: string
    /** Storage path, e.g. "jsonData.httpMethod" */
    path: string
    type: ValueType
    /** Human-readable field name */
    label?: string
    semanticType?: SemanticType
    description?: string
    required?: boolean
    requiredWhen?: string
    dependsOn?: string
    defaultValue?: unknown
    allowedValues?: unknown[]
    pattern?: string
    range?: { min?: number; max?: number }
    /** Nested fields for array/map items */
    items?: PromptField[]
    /** Virtual selector options with side-effects */
    options?: PromptOption[]
}

export interface PromptOption {
    value: unknown
    label: string
    sets: Record<string, unknown>
}

// ============================================================
// Projection
// ============================================================

/**
 * Projects a full DatasourceConfigSchema into a compact, LLM-friendly
 * PromptSchema by stripping UI hints, groups, storage mappings, and
 * other rendering/internal concerns.
 *
 * Keeps: identity, types, constraints, defaults, effects, descriptions.
 * Strips: ui, groups, overrides, lifecycle, tags, storage, isItemField.
 */
export function toPromptSchema(schema: DatasourceConfigSchema): PromptSchema {
    return {
        pluginType: schema.pluginType,
        pluginName: schema.pluginName,
        fields: schema.fields
            .filter((f) => !isManagedField(f))
            .map(projectField),
    }
}

/**
 * Serializes a DatasourceConfigSchema into a compact JSON string
 * suitable for embedding in an LLM prompt (tool/function calling).
 */
export function toPromptString(schema: DatasourceConfigSchema): string {
    return JSON.stringify(toPromptSchema(schema), null, 2)
}

/**
 * Renders a DatasourceConfigSchema into human-readable text suitable
 * for embedding in an LLM system/user prompt. More token-efficient
 * and easier for LLMs to reason about than JSON.
 *
 * Output format:
 * ```
 * Prometheus (pluginType: prometheus)
 *
 * Fields:
 * - url (root.url) [string, url] REQUIRED — Base URL
 * - httpMethod (jsonData.httpMethod) [string] default: "POST"
 *   Allowed: "GET", "POST"
 * ```
 */
export function toPromptText(schema: DatasourceConfigSchema): string {
    const ps = toPromptSchema(schema)
    const lines: string[] = []

    lines.push(`${ps.pluginName} (pluginType: ${ps.pluginType})`)
    lines.push("")
    lines.push("Fields:")

    for (const f of ps.fields) {
        lines.push(renderField(f, ""))
    }

    return lines.join("\n")
}

function renderField(f: PromptField, indent: string): string {
    const lines: string[] = []

    // Main line: - id (path) [type, semanticType] REQUIRED default: X — description
    const parts: string[] = []

    const name = f.label ?? f.id
    parts.push(`${indent}- ${name}`)

    // path (skip for virtual fields that have key-only path)
    if (f.path.includes(".")) {
        parts.push(`(${f.path})`)
    }

    // type annotation
    const typeAnnotations: string[] = [f.type]
    if (f.semanticType) typeAnnotations.push(f.semanticType)
    parts.push(`[${typeAnnotations.join(", ")}]`)

    if (f.required) parts.push("REQUIRED")
    if (f.defaultValue !== undefined) parts.push(`default: ${formatValue(f.defaultValue)}`)
    if (f.description) parts.push(`— ${f.description}`)

    lines.push(parts.join(" "))

    // Constraint lines (indented)
    const sub = indent + "  "

    if (f.requiredWhen) {
        lines.push(`${sub}Required when: ${f.requiredWhen}`)
    }
    if (f.dependsOn) {
        lines.push(`${sub}Visible when: ${f.dependsOn}`)
    }
    if (f.allowedValues && !f.options) {
        lines.push(`${sub}Allowed: ${f.allowedValues.map(formatValue).join(", ")}`)
    }
    if (f.pattern) {
        lines.push(`${sub}Pattern: ${f.pattern}`)
    }
    if (f.range) {
        const bounds: string[] = []
        if (f.range.min != null) bounds.push(`min: ${f.range.min}`)
        if (f.range.max != null) bounds.push(`max: ${f.range.max}`)
        lines.push(`${sub}Range: ${bounds.join(", ")}`)
    }

    // Options (virtual selector with effects)
    if (f.options) {
        lines.push(`${sub}Options:`)
        for (const opt of f.options) {
            const sets = Object.entries(opt.sets)
                .map(([k, v]) => `${k}=${formatValue(v)}`)
                .join(", ")
            lines.push(`${sub}  ${formatValue(opt.value)} (${opt.label}) → sets ${sets}`)
        }
    }

    // Array/map item fields
    if (f.items && f.items.length > 0) {
        lines.push(`${sub}Item fields:`)
        for (const item of f.items) {
            lines.push(renderField(item, sub + "  "))
        }
    }

    return lines.join("\n")
}

function formatValue(v: unknown): string {
    if (typeof v === "string") return `"${v}"`
    if (typeof v === "boolean" || typeof v === "number") return String(v)
    if (v === null || v === undefined) return "null"
    return JSON.stringify(v)
}

// ============================================================
// Internals
// ============================================================

/** Fields tagged "managed-by:*" are driven by effects and hidden from the LLM. */
function isManagedField(f: ConfigField): boolean {
    return f.tags?.some((t) => t.startsWith("managed-by:")) === true
}

function projectField(f: ConfigField): PromptField {
    const pf: PromptField = {
        id: f.id,
        path: fieldPath(f),
        type: f.valueType,
    }

    if (f.label) pf.label = f.label
    if (f.semanticType) pf.semanticType = f.semanticType
    if (f.description) pf.description = f.description
    if (f.required) pf.required = true
    if (f.requiredWhen) pf.requiredWhen = f.requiredWhen
    if (f.dependsOn) pf.dependsOn = f.dependsOn
    if (f.defaultValue !== undefined) pf.defaultValue = f.defaultValue

    // Flatten validations into top-level constraint fields
    if (f.validations) {
        for (const v of f.validations) {
            switch (v.type) {
                case "allowedValues":
                    pf.allowedValues = v.values
                    break
                case "pattern":
                    pf.pattern = v.pattern
                    break
                case "range":
                    pf.range = {}
                    if (v.min != null) pf.range.min = v.min
                    if (v.max != null) pf.range.max = v.max
                    break
            }
        }
    }

    // Flatten effects + ui.options into prompt options for virtual selectors
    if (f.effects && f.effects.length > 0) {
        const labelMap = buildLabelMap(f)
        pf.options = f.effects.map((eff) => {
            const val = extractLiteralFromWhen(eff.when)
            return {
                value: val ?? eff.when,
                label: labelMap.get(String(val)) ?? String(val),
                sets: eff.set,
            }
        })
    }

    // Recurse into array/map item fields
    if (f.item) {
        if (f.item.fields && f.item.fields.length > 0) {
            pf.items = f.item.fields.map(projectField)
        }
    }

    return pf
}

function fieldPath(f: ConfigField): string {
    if (!f.target) return f.key
    if (f.section) return `${f.target}.${f.section}.${f.key}`
    return `${f.target}.${f.key}`
}

/**
 * Builds a map from option value → label using ui.options.
 * Falls back to empty map if no UI options defined.
 */
function buildLabelMap(f: ConfigField): Map<string, string> {
    const m = new Map<string, string>()
    if (f.ui?.options) {
        for (const opt of f.ui.options) {
            m.set(String(opt.value), opt.label)
        }
    }
    return m
}

/**
 * Extracts the literal value from a simple `value == '...'` expression.
 * Returns undefined for complex expressions.
 */
export function extractLiteralFromWhen(when: string): unknown {
    // Match: value == 'string-literal'
    const singleQuote = /^value\s*==\s*'([^']*)'$/.exec(when)
    if (singleQuote) return singleQuote[1]

    // Match: value == "string-literal"
    const doubleQuote = /^value\s*==\s*"([^"]*)"$/.exec(when)
    if (doubleQuote) return doubleQuote[1]

    // Match: value == true / value == false
    const boolMatch = /^value\s*==\s*(true|false)$/.exec(when)
    if (boolMatch) return boolMatch[1] === "true"

    // Match: value == 123 (integer)
    const numMatch = /^value\s*==\s*(-?\d+(?:\.\d+)?)$/.exec(when)
    if (numMatch) return Number(numMatch[1])

    return undefined
}
